File: leetcode-go/Q051-Q100/Q052_N-Queens_II/Q052_N-Queens_II.go
package Q052_N_Queens_II

func totalNQueens(n int) int {
	board := make([][]byte, n)
	for i := range board {
		board[i] = make([]byte, n)
		for j := range board[i] {
			board[i][j] = '.'
		}
	}
	ans := 0
	dfs(board, 0, n, &ans)
	return ans
}

func dfs(board [][]byte, row, n int, ans *int) {
	if row == n {
		*ans++
	}
	for i := 0; i < n; i++ {
		if isValid(board, row, i, n) {
			board[row][i] = 'Q'
			dfs(board, row+1, n, ans)
			board[row][i] = '.'
		}
	}
}

func isValid(board [][]byte, x, y, n int) bool {
	for i := 0; i < x; i++ {
		if board[i][y] == 'Q' {
			return false
		}
	}
	for i, j := x-1, y-1; i >= 0 && j >= 0; {
		if board[i][j] == 'Q' {
			return false
		}
		i--
		j--
	}
	for i, j := x-1, y+1; i >= 0 && j < n; {
		if board[i][j] == 'Q' {
			return false
		}
		i--
		j++
	}
	return true
}

File: src/Y2018/backtracking/Q052_N_Queens_II.java
package Y2018.backtracking;

import java.util.HashSet;
import java.util.Set;

public class Q052_N_Queens_II {

    Set<Integer> cols = new HashSet<>();
    Set<Integer> diag = new HashSet<>();
    Set<Integer> antiDiag = new HashSet<>();
    int result = 0;

    public int totalNQueens(int n) {
        dfs(0, n);
        return result;
    }

    private void dfs(int row, int n) {
        if (row == n) {
            result++;
            return;
        }
        for (int col = 0; col < n; col++) {
            if (cols.contains(col)) {
                continue;
            }
            if (diag.contains(row + col)) {
                continue;
            }
            if (antiDiag.contains(row - col)) {
                continue;
            }
            cols.add(col);
            diag.add(row + col);
            antiDiag.add(row - col);
            dfs(row + 1, n);
            cols.remove(col);
            diag.remove(row + col);
            antiDiag.remove(row - col);
        }
    }

}
